### PR TITLE
ISPN-16823 Refactor RESP serialization

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.redhat.com/browse/ISPN-16823

Note to reviewers: this touches almost every class in the resp module, but is mostly just moving stuff around and encapsulating response writing to support the upcoming Lua scripting implementation which will require responses to be stored as objects rather than serialized to a byte buffer. As a bonus, it removes 66 lines of code :)

@jabolina could you please run some benchmarks to see if there are any performance regressions ?

### CODE REVIEW

**Comment by jabolina** [comment]
| # Branch   | avg      | min      | p50      | p95      | p99      | max       |
|------------|----------|----------|----------|----------|----------|-----------|
| main (GET) | 0.191    | 0.024    | 0.1723   | 0.3483   | 0.6763   | 15.2816   |
| This (GET) | 0.1903   | 0.0213   | 0.175    | 0.3483   | 0.655    | 13.1776   |
| main (SET) | 0.6293   | 0.456    | 0.615    | 0.719    | 1.327    | 10.9003   |
| this (SET) | 0.6063   | 0.4346   | 0.5963   | 0.6763   | 1.2763   | 11.8656   |
| Diff GET   | -0.349%  | -11.11%  | 1.5473%  | 0%       | -3.1542% | -13.7681% |
| Diff SET   | -3.6546% | -4.6783% | -3.0352% | -5.9341% | -3.8181% | 8.8555%   |


A negative value means the latency has reduced between versions. I'm going to review the code, but performance-wise, we're good.

**Review by jabolina — APPROVED** [review]
Looking good, just some small nits. Also, wondering if we should add something to separate writing error messages. So callers would do `writer.error().syntaxError()` or something like that, just to move this stuff from the `ResponseWriter` interface. But again, just nits.

File: server/resp/src/main/java/org/infinispan/server/resp/serialization/bytebuf/ByteBufPrimitiveSerializer.java [hunk 1]
```diff
@@ -35,19 +37,18 @@
  * @since 15.0
  * @author José Bolina
  */
-class PrimitiveSerializer {
-
+public class ByteBufPrimitiveSerializer {
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/resp/src/main/java/org/infinispan/server/resp/serialization/bytebuf/ByteBufPrimitiveSerializer.java:40]: sure

File: server/resp/src/main/java/org/infinispan/server/resp/serialization/NestedResponseSerializer.java [hunk 2]
```diff
@@ -16,11 +16,11 @@
  * @param <H> The generic type of hints the serializer accepts.
  * @author José Bolina
  */
-interface NestedResponseSerializer<T, H extends SerializationHint>
-      extends ResponseSerializer<T>, TriConsumer<T, ByteBufPool, H> {
+public interface NestedResponseSerializer<T, O, H extends SerializationHint>
+      extends ResponseSerializer<T, O>, TriConsumer<T, ByteBufPool, H> {
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply server/resp/src/main/java/org/infinispan/server/resp/serialization/NestedResponseSerializer.java:20]: right

**Comment by tristantarrant** [comment]
I addressed all reported issues aside from splitting the errors into another class, which, in my opinion, doesn't really help much

**Review by wburns — APPROVED** [review]
This approach seems good to me. The lua part will be much more interesting :)

**Comment by jabolina** [comment]
Thanks, @tristantarrant!

**Pull request merged**